bliss/models/location_encoder.py

```diff
@@ -216,55 +216,40 @@ def sample(self, var_params: Dict[str, Tensor], n_samples: int) -> Dict[str, Ten
-        pred = self._encode_for_n_sources(var_params["per_source_params"], tile_n_sources)
+        # get distributional parameters conditioned on the sampled numbers of light sources
+        pred = self._encode_for_n_sources(dist_params["per_source_params"], tile_n_sources)
 
         tile_is_on_array = get_is_on_from_n_sources(tile_n_sources, self.max_detections)
-        tile_is_on_array = tile_is_on_array.unsqueeze(-1).float()
-        pred["loc_sd"] = torch.exp(0.5 * pred["loc_logvar"])
-        pred["log_flux_sd"] = torch.exp(0.5 * pred["log_flux_logvar"])
-        tile_locs = self._sample_gated_normal(
-            pred["loc_mean"],
-            pred["loc_sd"],
-            rearrange(tile_is_on_array, "n b nth ntw ns 1 -> n (b nth ntw) ns 1"),
-        )
-        tile_log_fluxes = self._sample_gated_normal(
-            pred["log_flux_mean"],
-            pred["log_flux_sd"],
-            rearrange(tile_is_on_array, "n b nth ntw ns 1 -> n (b nth ntw) ns 1"),
-        )
-        # Why are we masking here?
-        tile_fluxes = tile_log_fluxes.exp() * rearrange(
-            tile_is_on_array, "n b nth ntw ns 1 -> n (b nth ntw) ns 1"
-        )
-        sample_flat = {
-            "locs": tile_locs,
-            "log_fluxes": tile_log_fluxes,
-            "fluxes": tile_fluxes,
-        }
-        b, nth, ntw, _ = log_probs_n_sources_per_tile.shape
-        sample = {}
-        for k, v in sample_flat.items():
-            pattern = "ns (b nth ntw) s k -> ns b nth ntw s k"
-            sample[k] = rearrange(v, pattern, b=b, nth=nth, ntw=ntw)
-        sample["n_sources"] = tile_n_sources
+        tile_is_on_array = rearrange(tile_is_on_array, "n b nth ntw ns -> n (b nth ntw) ns 1")
 
-        return sample
+        tile_locs = Normal(pred["loc_mean"], pred["loc_sd"]).rsample()
+        tile_locs *= tile_is_on_array  # Is masking here helpful/necessary?
+
+        tile_log_fluxes = Normal(pred["log_flux_mean"], pred["log_flux_sd"]).rsample()
+        tile_fluxes = tile_log_fluxes.exp()
+        tile_fluxes *= tile_is_on_array
+
+        # Given all the rearranging/unflattening below, I suspect we never should have
+        # flattened in the first place
+        b, nth, ntw, _ = dist_params["n_source_log_probs"].shape
+        unflatten = "ns (b nth ntw) s k -> ns b nth ntw s k"
+        return {
+            "locs": rearrange(tile_locs, unflatten, b=b, nth=nth, ntw=ntw),
+            "log_fluxes": rearrange(tile_log_fluxes, unflatten, b=b, nth=nth, ntw=ntw),
+            "fluxes": rearrange(tile_fluxes, unflatten, b=b, nth=nth, ntw=ntw),
+            "n_sources": tile_n_sources,
+        }
 
     def max_a_post(
-        self, var_params: Dict[str, Tensor], n_source_weights: Optional[Tensor] = None
+        self, dist_params: Dict[str, Tensor], n_source_weights: Optional[Tensor] = None
     ) -> TileCatalog:
-        """Derive maximum a posteriori from variational parameters.
+        """Compute the mode of the variational distribution.
```

Comment:
It's not exactly a map estimate.

Reply:
I agree - we maybe should rename the ".max_a_post()" methods to be ".mode()" instead, but that can be a separate PR

Reply:
Opened a new issue